cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java

```diff
@@ -375,6 +378,57 @@ public void deleteSchedules(Iterable<? extends ScheduleId> scheduleIds) throws N
   }
 
   /**
+   * Update all schedules that can be triggered by the given deleted program. Schedules will be removed if they
+   * contain single {@link co.cask.cdap.internal.app.runtime.schedule.trigger.ProgramStatusTrigger}. Schedules with
+   * composite triggers will be updated if the composite trigger can still be satisfied after the program is deleted,
+   * otherwise the schedules will be deleted.
+   *
+   * @param programId the program id for which to delete the schedules
+   * @return the IDs of the schedules that were deleted
+   */
+  public List<ScheduleId> modifySchedulesTriggeredByDeletedProgram(ProgramId programId) {
+    List<ScheduleId> deleted = new ArrayList<>();
+    Set<ProgramScheduleRecord> scheduleRecords = new HashSet<>();
+    for (ProgramStatus status : ProgramStatus.values()) {
+      scheduleRecords.addAll(findSchedules(Schedulers.triggerKeyForProgramStatus(programId, status)));
+    }
+    for (ProgramScheduleRecord scheduleRecord : scheduleRecords) {
+      ProgramSchedule schedule = scheduleRecord.getSchedule();
+      try {
+        deleteSchedule(schedule.getScheduleId());
```

Comment:
So the schedule is always deleted regardless of the trigger type??

Reply:
yes. If the schedule contains a valid composite after modification, the schedule will be added back. The update schedule method also deletes the schedule first and then add it back